socorro/schemas/validate_raw_crash.py

```diff
@@ -0,0 +1,202 @@
+    click.echo("socorro-data-1-0-0.schema.yaml is a valid jsonschema.")
+
+    jsonschema.validate(instance=RAW_CRASH_SCHEMA, schema=socorro_data_schema)
+    click.echo("raw crash schema is a valid socorro data schema.")
+
+    # Fetch crash report data from a Super Search URL
+    datapath = pathlib.Path(crashdir).resolve()
+    if not datapath.is_dir():
+        raise click.ClickException(f"{datapath} is not a directory.")
+
+    click.echo(f"Fetching data from {datapath}...")
+
+    uuids = list(datapath.glob("*"))
+
+    # Figure out the schema keys to types mapping
+    schema_key_logger = SchemaKeyLogger()
+    transform_schema(schema=RAW_CRASH_SCHEMA, transform_function=schema_key_logger)
+
+    schema_reducer = SocorroDataReducer(RAW_CRASH_SCHEMA)
+
+    document_keys = DocumentKeys()
+    reduced_keys = DocumentKeys()
+
+    total_uuids = len(uuids)
+    click.echo("")
+    click.echo(f"Testing {total_uuids} recent crash reports.")
+    for i, uuid in enumerate(uuids):
+        click.echo(f"Working on {uuid} ({i}/{total_uuids})...")
+        raw_crash = json.loads((datapath / uuid).read_text())
+
+        # Log the keys
+        document_keys.log_keys(raw_crash)
+
+        # Reduce the document by the schema and remove whatever keys are in the document
+        # which is what the schema knows about
+        reduced_raw_crash = schema_reducer.traverse(raw_crash)
+        reduced_keys.log_keys(reduced_raw_crash)
+
+        validate_instance(raw_crash, RAW_CRASH_SCHEMA)
+
+    click.echo("Done testing, all crash reports passed.")
+
+    reduced_keys_keys = set([key for key, type_ in reduced_keys.keys])
+
+    # Figure out which schema keys weren't in documents taking into account
+    # pattern_property regexes
+    keys_not_in_doc = set()
+    for key, type_ in schema_key_logger.keys:
+        is_in_reduced_keys = False
+        for reduced_key in reduced_keys_keys:
+            if match_key(key, reduced_key):
+                is_in_reduced_keys = True
+                break
+
+        if not is_in_reduced_keys:
+            keys_not_in_doc.add((key, type_))
+
+    if keys_not_in_doc:
+        click.echo(
+            f"{len(keys_not_in_doc)} (out of {len(schema_key_logger.keys)}) "
+            + "keys in JSON Schema, but never in any of the tested crashes:"
+        )
+        click.echo(f"  {'KEY':90}  TYPE(S)")
+        for key, val in sorted(keys_not_in_doc):
+            click.echo(f"  {key:90}  {val}")
+
+    # Figure out which doc keys aren't in the schema; this also handles cases where the
+    # key is in the schema, but is missing a type
+    keys_not_in_schema = document_keys.keys - reduced_keys.keys
+    if keys_not_in_schema:
+        click.echo("")
+        click.echo(
+            f"{len(keys_not_in_schema)} keys in crash reports but not in schema:"
+        )
+        click.echo(f"  {'KEY':90}  TYPE(S)")
+        for key, val in sorted(keys_not_in_schema):
+            click.echo(f"  {key:90}  {val}")
```

Comment:
It might be interesting to have this also fetch `CrashAnnotations.yaml` and then list out the annotations that are in the raw crash schema and not in `CrashAnnotations.yaml` and vice-versa.